app-directory/share-state/app/page.tsx

```diff
@@ -0,0 +1,178 @@
+import { Page, Text, Link, Code, Snippet } from '@vercel/examples-ui'
+
+export default function Home() {
+  return (
+    <Page className="flex flex-col gap-12">
+      <section className="flex flex-col gap-6">
+        <Text variant="h1">Share state within app dir</Text>
+        <Text>
+          Sometimes you need to share state between pages or layouts within your
+          app. It may be complex as your pages fetch data from the server or you
+          need to track state on your layout from user interactions within
+          nested routes.
+        </Text>
+      </section>
+
+      <section className="flex flex-col gap-3">
+        <Text variant="h2">Client layouts and server pages</Text>
+        <Text>
+          Althought client components can&apos;t include server components, we
+          can have a client layout rendering a page from the server. This allows
+          us to create a state or context to share data across our application.
+        </Text>
+      </section>
+
+      <section className="flex flex-col gap-3">
+        <Text variant="h2">Creating the files</Text>
+        <Text>
+          We will create a root layout for our app that will show the counter
+          state, a context to share state, an index and a nested page that
+          displays the counter component and a counter client component to
+          interact with our state.
+        </Text>
+        <pre className="border border-accents-2 rounded-md bg-accents-1 overflow-x-auto p-6">{`|/app
+|__/layout.js (client)
+|__/context.js (client)
+|__/counter.js (client)
+|__/page.js (server)
+|__/nested
+|____/page.js (server)
+`}</pre>
+        <Text>
+          Lets create a simple context that stores a <Code>counter</Code> value
+          and exports a <Code>useCounter</Code> hook to interact with that
+          value.
+        </Text>
+        <Snippet>
+          {`import { createContext, useContext, useState } from "react";
+
+const CounterContext = createContext({});
+
+function CounterProvider({ children }) {
+  const state = useState(0);
+
+  return <CounterContext.Provider value={state}>{children}</CounterContext.Provider>;
+}
+
+function useCounter() {
+  return useContext(CounterContext);
+}
+
+export { CounterProvider, useCounter };`}
```

Comment:
This use of useState is fine but it's not optimized for the use case that is being displayed in the demo, where the Layout **reads** and the pages **write**, because using the `useCounter` hook automatically subscribes you to both.

Here's a more optimized variant of this for a functionality that saves the current hash of a page and that can be modified for the counter:

```ts
import {
  useContext,
  useState,
  createContext,
  type Dispatch,
  type SetStateAction,
} from 'react';

type SetActiveHash = Dispatch<SetStateAction<string>>;

const ActiveHash = createContext('');
const SetActiveHash = createContext<SetActiveHash>(() => '');

export function ActiveHashProvider({
  children,
}: {
  children: React.ReactNode;
}): JSX.Element {
  const [activeHash, setActiveHash] = useState('');

  return (
    <ActiveHash.Provider value={activeHash}>
      <SetActiveHash.Provider value={setActiveHash}>
        {children}
      </SetActiveHash.Provider>
    </ActiveHash.Provider>
  );
}

export const useActiveHash = (): string => useContext(ActiveHash);

export const useSetActiveHash = (): SetActiveHash => useContext(SetActiveHash);
```

Reply:
I agree that its more performante, my concern is to complicate the example or make people question why there are two contexts :thonk:

Reply:
I think that can be solved with in-code comments, but having something like this is better for globally shared state like in this example.